File: gui/frontend/src/api/hra.ts
import { api } from './client'

// Human Reliability Analysis — one compute function per method. Each returns a
// human-error probability (HEP) plus method-specific detail.

// ── HEART ──
export interface HeartResponse {
  hep: number
  nominal: number
  gtt: string
  contributions: { epc_id: number; max_affect: number; proportion: number; factor: number; label: string }[]
}
export const computeHeart = (req: { gtt: string; epcs: { epc_id: number; proportion: number }[] }) =>
  api.post<HeartResponse>('/hra/heart', req).then(r => r.data)

// ── SPAR-H ──
export type SparHDependencyLevel = 'zero' | 'low' | 'moderate' | 'high' | 'complete'
export interface SparHDependencyRequest {
  enabled: boolean
  level?: SparHDependencyLevel | null
  same_crew?: boolean | null
  close_in_time?: boolean | null
  same_location?: boolean | null
  additional_cues?: boolean | null
  failure_number_in_sequence?: number
  justification?: string
}
export interface SparHResponse {
  hep: number
  nominal: number
  psf_product: number | null
  raw_hep?: number
  psf_adjusted_hep: number
  independent_hep: number
  adjustment_applied: boolean
  n_negative_psfs: number
  applied: Record<string, { level: string; multiplier: number | string }>
  guaranteed_failure: boolean
  minimum_cutoff_applied: boolean
  dependency: {
    applied: boolean
    level: SparHDependencyLevel
    independent_hep: number
    adjusted_hep: number
    source: string
    failure_number_in_sequence: number
    justification: string
    context?: Record<string, boolean | null>
  }
  uncertainty: {
    distribution: string
    mean: number
    alpha: number | null
    beta: number | null
    confidence: number
    lower: number
    upper: number
    median: number
    parameter_source: string
    note?: string
  }
  result_quality: 'screening'
  psf_dependence_note: string
}
export const computeSparH = (req: {
  task_type: string
  psfs: Record<string, string>
  dependency?: SparHDependencyRequest | null
  uncertainty_confidence?: number
  uncertainty_alpha?: number | null
}) =>
  api.post<SparHResponse>('/hra/spar-h', req).then(r => r.data)

// ── THERP ──
export interface TherpResponse {
  hep: number
  adjusted_hep: number
  nominal_hep: number
  stress_multiplier: number
  experience_multiplier: number
  conditional_hep: number | null
  joint_hep: number | null
  dependency: string | null
}
export const computeTherp = (req: {
  nominal_hep: number; stress: string; experience: string
  second_hep?: number | null; dependency?: string
}) => api.post<TherpResponse>('/hra/therp', req).then(r => r.data)

// ── CREAM ──
export interface CreamResponse {
  hep: number
  control_mode: string
  hep_lower: number
  hep_upper: number
  sum_reduced: number
  sum_improved: number
  effects: Record<string, string>
  /** Control-mode map: rows improved 0-7 × cols reduced 0-9; null = infeasible. */
  grid: (string | null)[][]
}
export const computeCream = (req: { cpc_levels: Record<string, string> }) =>
  api.post<CreamResponse>('/hra/cream', req).then(r => r.data)

// ── CREAM extended ──
export interface CreamExtStep {
  description: string
  activity: string
  failure_type: string
  failure_label: string
  function: string
  nominal_cfp: number
  weight: number
  cfp: number
}
export interface CreamExtendedResponse {
  hep: number
  steps: CreamExtStep[]
  dominant_step: CreamExtStep | null
  context_weights: Record<string, number>
}
export const computeCreamExtended = (req: {
  cpc_levels: Record<string, string>
  steps: { description: string; activity: string; failure_type: string }[]
}) => api.post<CreamExtendedResponse>('/hra/cream-extended', req).then(r => r.data)

// ── SLIM ──
export interface SlimResponse { hep: number; sli: number; a: number; b: number }
export const computeSlim = (req: {
  psfs: { weight: number; rating: number }[]
  anchors?: { sli: number; hep: number }[] | null
  a?: number | null; b?: number | null
}) => api.post<SlimResponse>('/hra/slim', req).then(r => r.data)

// ── Category-factor screening (legacy endpoint: JHEDI) ──
export interface CategoryScreeningResponse {
  hep: number
  base: number
  aggravating_factors: number
  factor_multiplier: number
  result_quality: 'screening'
  warning: string
}
export type JhediResponse = CategoryScreeningResponse
export const computeCategoryScreening = (req: { task_category: string; aggravating_factors: number }) =>
  api.post<CategoryScreeningResponse>('/hra/category-screening', req).then(r => r.data)
export const computeJhedi = (req: { task_category: string; aggravating_factors: number }) =>
  api.post<JhediResponse>('/hra/jhedi', req).then(r => r.data)

// ── Error-mode likelihood screening (SHERPA-inspired taxonomy) ──
export interface ErrorModeScreeningResponse {
  hep: number
  overall_error_probability: number
  max_critical_probability: number
  counts_by_mode: Record<string, number>
  rows: { error_mode: string; probability: number; critical: boolean }[]
  result_quality: 'screening'
  assumption: string
}
export type SherpaResponse = ErrorModeScreeningResponse
export const computeErrorModeScreening = (req: {
  rows: { error_mode: string; probability: string; critical: boolean }[]
}) => api.post<ErrorModeScreeningResponse>('/hra/error-mode-screening', req).then(r => r.data)
export const computeSherpa = (req: {
  rows: { error_mode: string; probability: string; critical: boolean }[]
}) => api.post<SherpaResponse>('/hra/sherpa', req).then(r => r.data)

// ── Error-forcing-context elicitation screen (not full ATHEANA) ──
export interface EfcElicitationResponse {
  hep: number
  min: number
  mode: number
  max: number
  result_quality: 'screening'
  warning: string
}
export type AtheanaResponse = EfcElicitationResponse
export const computeEfcElicitationScreening = (req: { min_hep: number; mode_hep: number; max_hep: number }) =>
  api.post<EfcElicitationResponse>('/hra/efc-elicitation-screening', req).then(r => r.data)
export const computeAtheana = (req: { min_hep: number; mode_hep: number; max_hep: number }) =>
  api.post<AtheanaResponse>('/hra/atheana', req).then(r => r.data)

// ── Mission-scenario screening sum (not full MERMOS) ──
export interface MissionScenarioScreeningResponse {
  hep: number
  total_failure_probability: number
  dominant_scenario: { label: string; probability: number } | null
  scenarios: { label: string; probability: number }[]
  mutually_exclusive: boolean
  result_quality: 'screening'
  warning: string
}
export type MermosResponse = MissionScenarioScreeningResponse
export const computeMissionScenarioScreening = (req: {
  scenarios: { label: string; probability: number }[]
  mutually_exclusive: boolean
}) => api.post<MissionScenarioScreeningResponse>('/hra/mission-scenario-screening', req).then(r => r.data)
export const computeMermos = (req: { scenarios: { label: string; probability: number }[] }) =>
  api.post<MermosResponse>('/hra/mermos', req).then(r => r.data)

File: gui/backend/routers/hra.py
"""Human Reliability Analysis (HRA) router.

Exposes the implemented HRA methods in reliability.HRA plus clearly identified
screening worksheets. Legacy JHEDI/SHERPA/ATHEANA/MERMOS paths remain available
for saved clients, but return warnings because they are not full implementations
of those published methods. Bad input raises ValueError → HTTP 400 via the
global handler in main.py.
"""

import sys
from pathlib import Path
from typing import Dict, List, Literal, Optional

from fastapi import APIRouter
from pydantic import BaseModel, Field

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "src"))

from reliability import HRA

from utils import safe as _safe

router = APIRouter()


# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------

class EpcItem(BaseModel):
    epc_id: int
    proportion: float = Field(..., ge=0, le=1)


class HeartRequest(BaseModel):
    gtt: str
    epcs: List[EpcItem] = []


class SparHDependency(BaseModel):
    enabled: bool = False
    level: Optional[Literal['zero', 'low', 'moderate', 'high', 'complete']] = None
    same_crew: Optional[bool] = None
    close_in_time: Optional[bool] = None
    same_location: Optional[bool] = None
    additional_cues: Optional[bool] = None
    failure_number_in_sequence: int = Field(2, ge=2)
    justification: str = ""


class SparHRequest(BaseModel):
    task_type: str = "action"
    psfs: Dict[str, str] = {}
    dependency: Optional[SparHDependency] = None
    uncertainty_confidence: float = Field(0.90, gt=0, lt=1)
    uncertainty_alpha: Optional[float] = Field(None, gt=0)


class TherpRequest(BaseModel):
    nominal_hep: float = Field(..., ge=0, le=1)
    stress: str = "optimal"
    experience: str = "skilled"
    second_hep: Optional[float] = Field(None, ge=0, le=1)
    dependency: str = "ZD"


class CreamRequest(BaseModel):
    cpc_levels: Dict[str, str] = {}


class CreamStep(BaseModel):
    description: str = ""
    activity: str
    failure_type: str


class CreamExtendedRequest(BaseModel):
    cpc_levels: Dict[str, str] = {}
    steps: List[CreamStep]


class SlimPsf(BaseModel):
    weight: float = Field(..., ge=0)
    rating: float


class SlimAnchor(BaseModel):
    sli: float
    hep: float = Field(..., gt=0, lt=1)


class SlimRequest(BaseModel):
    psfs: List[SlimPsf]
    anchors: Optional[List[SlimAnchor]] = None
    a: Optional[float] = None
    b: Optional[float] = None


class JhediRequest(BaseModel):
    task_category: str
    aggravating_factors: int = Field(0, ge=0)


class SherpaRow(BaseModel):
    error_mode: str = "unspecified"
    probability: str = "M"
    critical: bool = False


class SherpaRequest(BaseModel):
    rows: List[SherpaRow]


class AtheanaRequest(BaseModel):
    min_hep: float = Field(..., ge=0, le=1)
    mode_hep: float = Field(..., ge=0, le=1)
    max_hep: float = Field(..., ge=0, le=1)


class MermosScenario(BaseModel):
    label: str = "scenario"
    probability: float = Field(..., ge=0, le=1)


class MermosRequest(BaseModel):
    scenarios: List[MermosScenario]
    mutually_exclusive: bool = False


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------

@router.post("/heart")
def heart(req: HeartRequest):
    return _safe(HRA.heart(req.gtt, [e.model_dump() for e in req.epcs]))


@router.post("/spar-h")
def spar_h(req: SparHRequest):
    return _safe(HRA.spar_h(
        req.task_type,
        req.psfs,
        dependency=req.dependency.model_dump() if req.dependency else None,
        uncertainty_confidence=req.uncertainty_confidence,
        uncertainty_alpha=req.uncertainty_alpha,
    ))


@router.post("/therp")
def therp(req: TherpRequest):
    return _safe(HRA.therp(req.nominal_hep, req.stress, req.experience,
                           req.second_hep, req.dependency))


@router.post("/cream")
def cream(req: CreamRequest):
    return _safe(HRA.cream(req.cpc_levels))


@router.post("/cream-extended")
def cream_extended(req: CreamExtendedRequest):
    return _safe(HRA.cream_extended(req.cpc_levels, [s.model_dump() for s in req.steps]))


@router.post("/slim")
def slim(req: SlimRequest):
    return _safe(HRA.slim([p.model_dump() for p in req.psfs],
                          [a.model_dump() for a in req.anchors] if req.anchors else None,
                          req.a, req.b))


@router.post("/jhedi")
def jhedi(req: JhediRequest):
    return _safe(HRA.jhedi(req.task_category, req.aggravating_factors))


@router.post("/category-screening")
def category_screening(req: JhediRequest):
    return _safe(HRA.category_factor_screening(
        req.task_category, req.aggravating_factors))


@router.post("/sherpa")
def sherpa(req: SherpaRequest):
    return _safe(HRA.sherpa([r.model_dump() for r in req.rows]))


@router.post("/error-mode-screening")
def error_mode_screening(req: SherpaRequest):
    return _safe(HRA.error_mode_screening([r.model_dump() for r in req.rows]))


@router.post("/atheana")
def atheana(req: AtheanaRequest):
    return _safe(HRA.atheana(req.min_hep, req.mode_hep, req.max_hep))


@router.post("/efc-elicitation-screening")
def efc_elicitation_screening(req: AtheanaRequest):
    return _safe(HRA.efc_elicitation_screening(
        req.min_hep, req.mode_hep, req.max_hep))


@router.post("/mermos")
def mermos(req: MermosRequest):
    return _safe(HRA.mermos([s.model_dump() for s in req.scenarios]))


@router.post("/mission-scenario-screening")
def mission_scenario_screening(req: MermosRequest):
    return _safe(HRA.mission_scenario_screening(
        [s.model_dump() for s in req.scenarios], req.mutually_exclusive))
